carma_cooperative_perception/src/msg_conversion.cpp

```diff
@@ -55,18 +57,74 @@
 
 namespace carma_cooperative_perception
 {
-auto to_time_msg(const DDateTime & d_date_time) -> builtin_interfaces::msg::Time
+auto to_time_msg(const DDateTime & d_date_time, bool is_simulation) -> builtin_interfaces::msg::Time
 {
-  double seconds;
-  const auto fractional_secs{std::modf(
-    remove_units(units::time::second_t{d_date_time.hour.value_or(units::time::second_t{0.0})}) +
-      remove_units(units::time::second_t{d_date_time.minute.value_or(units::time::second_t{0.0})}) +
-      remove_units(units::time::second_t{d_date_time.second.value_or(units::time::second_t{0.0})}),
-    &seconds)};
-
+  // Convert DDateTime to builtin_interfaces::msg::Time
   builtin_interfaces::msg::Time msg;
-  msg.sec = static_cast<std::int32_t>(seconds);
-  msg.nanosec = static_cast<std::int32_t>(fractional_secs * 1e9);
+  if (!is_simulation) {
+    // Create a tm structure to hold the date and time components
+    std::tm timeinfo = {};
+    timeinfo.tm_year = static_cast<int>(d_date_time.year.value());
+    if(timeinfo.tm_year > 1900){
+        timeinfo.tm_year = timeinfo.tm_year - 1900;
+    }
+    timeinfo.tm_mon = static_cast<int>(d_date_time.month.value().get_value());
+    if (timeinfo.tm_mon != 0){
+        timeinfo.tm_mon = timeinfo.tm_mon - 1;
+    }
+    timeinfo.tm_mday = static_cast<int>(d_date_time.day.value());
+    timeinfo.tm_hour = static_cast<int>(d_date_time.hour.value());
+    timeinfo.tm_min = static_cast<int>(d_date_time.minute.value());
+    timeinfo.tm_sec = 0;
+    timeinfo.tm_isdst = 1;  // Force EDT (Daylight Saving Time)
+
+    // Store current TZ environment variable
+    char* old_tz = getenv("TZ");
+    std::string old_tz_str = old_tz ? old_tz : "";
+
+    // Set timezone to EDT
+    setenv("TZ", "America/New_York", 1);
+    tzset();
+
+    // Convert EDT time to time_t (which is always in UTC)
+    std::time_t timeT = std::mktime(&timeinfo);
+
+    // Restore original timezone
+    if (old_tz_str.empty()) {
+        unsetenv("TZ");
+    } else {
+        setenv("TZ", old_tz_str.c_str(), 1);
+    }
+    tzset();
+
+    // Convert time_t to system_clock::time_point
+    auto timePoint = std::chrono::system_clock::from_time_t(timeT);
+
+    // Add milliseconds
+    auto milliseconds = static_cast<int>(d_date_time.second.value());
+    timePoint += std::chrono::milliseconds(milliseconds);
+
+    // Extract seconds and nanoseconds since epoch
+    auto duration = timePoint.time_since_epoch();
+    auto seconds = std::chrono::duration_cast<std::chrono::seconds>(duration);
+    auto nanoseconds = std::chrono::duration_cast<std::chrono::nanoseconds>(duration - seconds);
+
+    msg.sec = static_cast<int32_t>(seconds.count());
+    msg.nanosec = static_cast<uint32_t>(nanoseconds.count());
```

Comment:
This is just one way of converting the timestamps, other code uses boost. Please let know if we want different logic. This way needs to hardcode the timezone representation to EDT/EST to correctly interpret otherwise it always reports it back in GMT.

Reply:
As discussed in person, since this implementation has been shown to work reliably in vehicle and simulation we will keep this.